packages/react-query-next-experimental/src/HydrationStreamProvider.tsx

```diff
@@ -0,0 +1,192 @@
+'use client'
+
+import { useServerInsertedHTML } from 'next/navigation'
+import * as React from 'react'
+
+const serializedSymbol = Symbol('serialized')
+
+interface DataTransformer {
+  serialize(object: any): any
+  deserialize(object: any): any
+}
+
+type Serialized<TData> = unknown & {
+  [serializedSymbol]: TData
+}
+
+interface TypedDataTransformer<TData> {
+  serialize: (obj: TData) => Serialized<TData>
+  deserialize: (obj: Serialized<TData>) => TData
+}
+
+interface HydrationStreamContext<TShape> {
+  id: string
+  stream: {
+    /**
+     * **Server method**
+     * Push a new entry to the stream
+     * Will be ignored on the client
+     */
+    push: (...shape: TShape[]) => void
+  }
+}
+
+export interface HydrationStreamProviderProps<TShape> {
+  children: React.ReactNode
+  /**
+   * Optional transformer to serialize/deserialize the data
+   * Example devalue, superjson et al
+   */
+  transformer?: DataTransformer
+  /**
+   * **Client method**
+   * Called in the browser when new entries are received
+   */
+  onEntries: (entries: TShape[]) => void
+  /**
+   * **Server method**
+   * onFlush is called on the server when the cache is flushed
+   */
+  onFlush?: () => TShape[]
+}
+
+export function createHydrationStreamProvider<TShape>() {
+  const context = React.createContext<HydrationStreamContext<TShape>>(null as any)
+  /**
+
+   * 1. (Happens on server): `useServerInsertedHTML()` is called **on the server** whenever a `Suspense`-boundary completes
+   *    - This means that we might have some new entries in the cache that needs to be flushed
+   *    - We pass these to the client by inserting a `<script>`-tag where we do `window[id].push(serializedVersionOfCache)`
+   * 2. (Happens in browser) In `useEffect()`:
+   *   - We check if `window[id]` is set to an array and call `push()` on all the entries which will call `onEntries()` with the new entries
+   *   - We replace `window[id]` with a `push()`-method that will be called whenever new entries are received
+   **/
+  function UseClientHydrationStreamProvider(props: {
+    children: React.ReactNode
+    /**
+     * Optional transformer to serialize/deserialize the data
+     * Example devalue, superjson et al
+     */
+    transformer?: DataTransformer
+    /**
+     * **Client method**
+     * Called in the browser when new entries are received
+     */
+    onEntries: (entries: TShape[]) => void
+    /**
+     * **Server method**
+     * onFlush is called on the server when the cache is flushed
+     */
+    onFlush?: () => TShape[]
+  }) {
+    // unique id for the cache provider
+    const id = `_${React.useId()}`
+    const idJSON = JSON.stringify(id)
```

Comment:
since we put `idJSON` on the `window`, should we prefix it with something static like `RQ_` ?

Reply:
Yeah, I think react/next uses to underscores and a namespace

Originally we had a fixed namespace like `__cache` where I added stuff, like the structure was `window.__cache[useId()] = []`

Reply:
namespaced here: fede312c